sharding/contracts/sharding_manager_test.go

```diff
@@ -1,130 +1,560 @@
+			"Got - deposited:%v, index:%v, period:%v ", notary.Deposited, notary.PoolIndex, notary.DeregisteredPeriod)
+	}
+
+	notary, err = smc.NotaryRegistry(&bind.CallOpts{}, notaryPoolAddr[2])
+
+	if !notary.Deposited &&
+		notary.PoolIndex.Cmp(big.NewInt(2)) != 0 &&
+		notary.DeregisteredPeriod.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect notary registry. Want - deposited:true, index:2, period:0"+
+			"Got - deposited:%v, index:%v, period:%v ", notary.Deposited, notary.PoolIndex, notary.DeregisteredPeriod)
+	}
+
+	// Check total numbers of notaries in pool
+	numNotaries, err := smc.NotaryPoolLength(&bind.CallOpts{})
 	if err != nil {
-		t.Fatalf("Error getting collationGasLimit: %v", err)
+		t.Fatalf("Failed to get notary pool length: %v", err)
 	}
-	if gasLimit.Cmp(big.NewInt(10000000)) != 0 {
-		t.Fatalf("collation gas limit should be 10000000 gas")
+	if numNotaries.Cmp(big.NewInt(3)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 3, Got: %v", numNotaries)
 	}
 }
 
-// Test collator deposit
-func TestCollatorDeposit(t *testing.T) {
-	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance1001Eth}})
-	transactOpts := bind.NewKeyedTransactor(key)
-	_, _, smc, _ := deploySMCContract(backend)
+func TestNotaryRegisterInsufficientEther(t *testing.T) {
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	txOpts := bind.NewKeyedTransactor(mainKey)
+	txOpts.Value = notaryDepositInsufficient
+	_, _, smc, _ := deploySMCContract(backend, mainKey)
+
+	smc.RegisterNotary(txOpts)
+	backend.Commit()
+
+	notary, _ := smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
+
+	if notary.Deposited {
+		t.Fatalf("Notary deposited with insufficient fund")
+	}
+
+	if numNotaries.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 0, Got: %v", numNotaries)
+	}
+}
+
+func TestNotaryDoubleRegisters(t *testing.T) {
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	txOpts := bind.NewKeyedTransactor(mainKey)
+	txOpts.Value = notaryDeposit
+	_, _, smc, _ := deploySMCContract(backend, mainKey)
 
-	// Test deposit() function
-	// Deposit 100 Eth
-	transactOpts.Value = collatorDeposit
+	// Notary 0 registers
+	smc.RegisterNotary(txOpts)
+	backend.Commit()
+
+	notary, _ := smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
+
+	if !notary.Deposited {
+		t.Fatalf("Notary has not registered. Got deposited flag: %v", notary.Deposited)
+	}
+
+	if numNotaries.Cmp(big.NewInt(1)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 1, Got: %v", numNotaries)
+	}
+
+	// Notary 0 registers again
+	_, _ = smc.RegisterNotary(txOpts)
```

Comment:
Can we verify the transaction failed and there was no state change? i.e. the sender did not change some variables in the contract and did not lose their 2000 ETH but double depositing

Reply:
Sure. I'll add additional check